x-pack/legacy/plugins/epm/server/datasources/create.ts

```diff
@@ -107,40 +85,33 @@ async function baseSetup(callCluster: CallESAsCurrentUser) {
   }
 }
 
-async function saveDatasourceReferences(options: {
+async function createDatasourceObject(options: {
   savedObjectsClient: SavedObjectsClientContract;
   pkg: RegistryPackage;
-  datasets: Dataset[];
   datasourceName: string;
   toSave: AssetReference[];
-  request: Request;
-  streams: Stream[];
+  datasets: Dataset[];
 }) {
-  const { savedObjectsClient, pkg, toSave, datasets, datasourceName, request, streams } = options;
-  const savedDatasource = await getDatasource({ savedObjectsClient, pkg });
-  const savedAssets = savedDatasource?.package.assets;
-  const assetsReducer = (current: Asset[] = [], pending: Asset) => {
-    const hasAsset = current.find(c => c.id === pending.id && c.type === pending.type);
-    if (!hasAsset) current.push(pending);
-    return current;
-  };
+  const { savedObjectsClient, pkg, toSave, datasets, datasourceName } = options;
+  const savedDatasource = await getDatasource({ savedObjectsClient, name: datasourceName });
```

Comment:
Find the datasource by name; not the id

Reply:
Can you share some background on what it is searched by name instead of id?

Reply:
This came about during a discussion in Slack and the weekly sync meeting Dec 19.

EPM was including an id in the datasource object when POSTing to Ingest API. Ingest was ignoring the id from the payload and letting Elasticsearch generate one. Ingest fixed the bug which ignored the id, but while discussing it,  we realized a) EPM was supplying the wrong id value ('pkgname-pkgversion' instead of dataset.name; preventing multiple sources per package) and b) EPM didn't really need to "know" the id; it could store a reference like it does with other assets.

Since dataset.name is supposed to be unique we use that to search for any existing instance of the datasource. This way there is a predictable way to find existing datasources (that doesn't fail with multiple datasources per package like `pkgname-pkgversion` does) and Ingest / ES can manage the IDs (which is more durable because they can be updated without any need to coordinated with EPM).